### DISCUSSION TITLE
Deadband filtering for server clients

### DISCUSSION THREAD

**Discussion by dakotahorstman** [FreeOpcUa/opcua-asyncio#1640] [Q&A]
Using Server, how can I deadband filter data changes before they are propagated to clients?

For example, [in Ignition](https://www.docs.inductiveautomation.com/docs/8.1/platform/tags/tag-properties#numeric-properties), it's possible to specify a deadband threshold and type where tag changes will not be propagated unless they exceed this threshold. Specificially, the Ignition server will receive a data change, inspect if its change %age or value exceeds the configured deadband, and, if it did, then inform the other clients of that change. If it did not, the server ignores the change. Looking at the source code for asyncua, I can see [deadband filtering is supported](https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua/server/monitored_item_service.py#L227), but can it be used in the way I'm requesting?

I believe this transcends the `create_subscription` `deadband_monitor` mentioned [here](https://github.com/FreeOpcUa/opcua-asyncio/discussions/1280) as I'm looking to filter changes before they are propagated to clients, not just subscriptions I create.

The documentation is heavily lacking in this area (I understand it is a WIP).

EDIT: "clients" refer to non-asyncua devices and programs such as Beckhoff Twincat and UAExpert. I am hosting an asyncua OPC Server and connecting to it from external devices.

**Comment by AndreasHeine** [comment]
subscription class has a method to create the monitoreditem with filters https://github.com/FreeOpcUa/opcua-asyncio/blob/master/asyncua%2Fcommon%2Fsubscription.py#L450

**Reply by dakotahorstman** [reply]
Not sure I follow. I'm aware of the method, but I can modify existing client session subscriptions with it? Assuming that's true, how do I get their subscriptions to modify then?

**Reply by AndreasHeine** [reply]
the client has to modify its subscription!

**Reply by dakotahorstman** [reply]
Ah. You're misunderstanding my questions. I am hosting an OPCUA server with asyncua and connecting to it via Beckhoff Twincat or UAExpert. I am not creating the clients with asyncua nor do I have the ability to configure this on their end.

**Reply by AndreasHeine** [reply]
the whole idea in opc ua subscription's is to give control to the client in order to not overwhelm clients on initialisation! if you are the server you should not touch the clients subscription!!!!!!!!!

the only thing you can control is the way the variables got written!

**Reply by dakotahorstman** [reply]
Got it.

EDIT:

Beckhoff Twincat does not support client-side filtering as expected by the OPC spec. At least not without a lot of structured text.

@AndreasHeine Due to this confirmation, having a pointer to implement this on the server's end would be great. I understand it differs from the spec. I understand asyncua wants to be a purist and adhere strictly to the spec. However, for practical reasons, I cannot. Hence why Ignition also implemented server-side deadband filtering--not all vendors may support it natively.

I'm not asking for a rewrite of core internals of asyncua by the maintainers, I am just asking for pointers and help on how to do this myself.

**Reply by AndreasHeine** [reply]
just let me get it straight first:
you are using asyncua server and beckhoff plc is client?

**Reply by dakotahorstman** [reply]
That is correct, yes. Beckhoff support confirmed there is no natively supported OPC filtering by Twincat. It would have to be implemented manually via Structured Text. This is a challenge as I have _a lot_ of Twincat PLCs.

**Reply by AndreasHeine** [reply]
so you need to intercept the writing of the variables the beckhoff client subscribes!

DataSource -> OPC UA Server -> OPC UA Client

![image007](https://github.com/FreeOpcUa/opcua-asyncio/assets/56362817/00bb34f7-1b0f-4cd5-910e-2cb96266fd05)
Source: https://reference.opcfoundation.org/Core/Part1/v105/docs/6.3

```python
async def write_variable_deadband(node: Node, dv: ua.DataValue):
    # do your deadbeand calc here
    if outside_deadband:
        await node.write_attribute(ua.AttributeIds.Value, dv)
# node.write_value would work as well but you'll pass a Variant instead of a DataValue 
```
